File: lista07/BinaryTree.cpp
// File: BinaryTree.cpp

#include <iostream>
#include <memory>
#include <vector>
#include <string>
#include <algorithm>

/**
 * A generic binary tree node.
 *
 * @tparam T the type of value stored in the node
 */
template <typename T>
class TreeNode {
public:
    T value;
    std::shared_ptr<TreeNode<T>> left;
    std::shared_ptr<TreeNode<T>> right;

    TreeNode(T val) : value(val), left(nullptr), right(nullptr) {}
};

/**
 * A generic binary tree implementation.
 *
 * @tparam T the type of value stored in the tree
 */
template <typename T>
class BinaryTree {
private:
    std::shared_ptr<TreeNode<T>> root;

    std::shared_ptr<TreeNode<T>> insertRec(std::shared_ptr<TreeNode<T>> node, T value) {
        if (!node) return std::make_shared<TreeNode<T>>(value);
        if (value < node->value) node->left = insertRec(node->left, value);
        else if (value > node->value) node->right = insertRec(node->right, value);
        return node;
    }

    std::shared_ptr<TreeNode<T>> deleteRec(std::shared_ptr<TreeNode<T>> node, T value) {
        if (!node) return node;

        if (value < node->value) node->left = deleteRec(node->left, value);
        else if (value > node->value) node->right = deleteRec(node->right, value);
        else {
            if (!node->left) return node->right;
            if (!node->right) return node->left;

            node->value = minValue(node->right);
            node->right = deleteRec(node->right, node->value);
        }
        return node;
    }

    T minValue(std::shared_ptr<TreeNode<T>> node) {
        T minVal = node->value;
        while (node->left) {
            minVal = node->left->value;
            node = node->left;
        }
        return minVal;
    }

    bool searchRec(std::shared_ptr<TreeNode<T>> node, T value) {
        if (!node) return false;
        if (node->value == value) return true;
        if (value < node->value) return searchRec(node->left, value);
        return searchRec(node->right, value);
    }

    void drawRec(std::shared_ptr<TreeNode<T>> node, int space) {
        if (!node) return;

        space += 10;
        drawRec(node->right, space);

        std::cout << std::endl;
        for (int i = 10; i < space; i++) std::cout << " ";
        std::cout << node->value << "\n";

        drawRec(node->left, space);
    }

public:
    void insert(T value) {
        root = insertRec(root, value);
    }

    void deleteNode(T value) {
        root = deleteRec(root, value);
    }

    bool search(T value) {
        return searchRec(root, value);
    }

    void draw() {
        drawRec(root, 0);
    }
};

int main() {
    BinaryTree<int> intTree;
    intTree.insert(50);
    intTree.insert(30);
    intTree.insert(70);
    intTree.insert(20);
    intTree.insert(40);
    intTree.insert(60);
    intTree.insert(80);
    intTree.draw();
    std::cout << "Drzewo po zmianach\n";
    intTree.deleteNode(20);
    intTree.deleteNode(30);
    intTree.draw();
    intTree.search(30) ? std::cout << "30 istnieje\n" : std::cout << "30 nie istnieje\n";
    intTree.search(40) ? std::cout << "40 istnieje\n" : std::cout << "40 nie istnieje\n";
    std::cout << "\n";

    BinaryTree<std::string> stringTree;
    stringTree.insert("M");
    stringTree.insert("B");
    stringTree.insert("Q");
    stringTree.insert("A");
    stringTree.insert("C");
    stringTree.insert("O");
    stringTree.insert("R");
    stringTree.draw();
     std::cout << "Drzewo po zmianach\n";
    stringTree.deleteNode("A");
    stringTree.deleteNode("R");
    stringTree.draw();
    stringTree.search("R") ? std::cout << "R istnieje\n" : std::cout << "R nie istnieje\n";
    stringTree.search("C") ? std::cout << "C istnieje\n" : std::cout << "C nie istnieje\n";
    std::cout << "\n";
    BinaryTree<double> doubleTree;
    doubleTree.insert(50.5);
    doubleTree.insert(30.3);
    doubleTree.insert(70.2);
    doubleTree.insert(20.1);
    doubleTree.insert(40.5);
    doubleTree.insert(60.4);
    doubleTree.insert(80.6);
    doubleTree.draw();
     std::cout << "Drzewo po zmianach\n";
    doubleTree.deleteNode(20.1);
    doubleTree.deleteNode(60.4);
    doubleTree.draw();
    doubleTree.search(60.4) ? std::cout << "60.4 istnieje\n" : std::cout << "60.4 nie istnieje\n";
    doubleTree.search(40.5) ? std::cout << "40.5 istnieje\n" : std::cout << "40.5 nie istnieje\n";
    std::cout << "\n";

    return 0;
}

File: lista07/Java/BinaryTree.java
/* Plik: BinaryTree.java
* Autor: Rafał Wochna 279752
*/
import java.io.Serializable;
/**
 * Drzewo binarne.
 * @param <T> typ wartości w drzewie
 */
class TreeNode<T extends Comparable<T>> implements Serializable {
    T value; // Wartość węzła.
    TreeNode<T> left; /* Lewe dziecko węzła.*/ 
    TreeNode<T> right; /* Prawe dziecko węzła.*/ 

    TreeNode(T value) {
        this.value = value; /* Ustaw wartość węzła. */
        this.left = null; /* Ustaw lewe dziecko na null. */
        this.right = null; /* Ustaw prawe dziecko na null. */
    } 
}
/**
 * Drzewo binarne.
 * @param <T> typ wartości w drzewie
 * @see TreeNode
 * @see Serializable
 * @see Comparable
 */
public class BinaryTree<T extends Comparable<T>> implements Serializable {
    private TreeNode<T> root; /* Korzeń drzewa. */
/**
 * Wstawia wartość do drzewa.
 * @param value - wartość do wstawienia
 */
    public void insert(T value) {
        root = insertRec(root, value);
    }
/**
 * Wstawia wartość do drzewa przy użyciu rekurencji.
 * @param root - korzeń drzewa
 * @param value - wartość do wstawienia
 * @return korzeń drzewa
 */
    private TreeNode<T> insertRec(TreeNode<T> root, T value) {
        if (root == null) {
            root = new TreeNode<>(value);
            return root;
        }
        if (value.compareTo(root.value) < 0) {
            root.left = insertRec(root.left, value);
        } else if (value.compareTo(root.value) > 0) {
            root.right = insertRec(root.right, value);
        }
        return root;
    }
/**
 * Usuwa wartość z drzewa.
 * @param value - wartość do usunięcia
 */
    public void delete(T value) {
        root = deleteRec(root, value);
    }
/**
 * Usuwa wartość z drzewa przy użyciu rekurencji.
 * @param root - korzeń drzewa
 * @param value - wartość do usunięcia
 * @return korzeń drzewa
 */
    private TreeNode<T> deleteRec(TreeNode<T> root, T value) {
        if (root == null) return root;

        if (value.compareTo(root.value) < 0) {
            root.left = deleteRec(root.left, value);
        } else if (value.compareTo(root.value) > 0) {
            root.right = deleteRec(root.right, value);
        } else {
            if (root.left == null) return root.right;
            if (root.right == null) return root.left;

            root.value = minValue(root.right);
            root.right = deleteRec(root.right, root.value);
        }

        return root;
    }
/**
 * Znajduje najmniejszą wartość w drzewie.
 * @param root - korzeń drzewa
 * @return najmniejsza wartość w drzewie
 */
    private T minValue(TreeNode<T> root) {
        T minVal = root.value;
        while (root.left != null) {
            minVal = root.left.value;
            root = root.left;
        }
        return minVal;
    }
/**
 * Wyszukuje wartość w drzewie.
 * @param value - wartość do wyszukania
 * @return true, jeśli wartość zostanie znaleziona, false w przeciwnym razie
 */
    public boolean search(T value) {
        return searchRec(root, value) != null;
    }
/**
 * Wyszukuje wartość w drzewie przy użyciu rekurencji.
 * @param root - korzeń drzewa
 * @param value - wartość do wyszukania
 * @return węzeł zawierający wartość lub null, jeśli wartość nie zostanie znaleziona
 */
    private TreeNode<T> searchRec(TreeNode<T> root, T value) {
        if (root == null || root.value.equals(value)) {
            return root;
        }
        if (value.compareTo(root.value) < 0) {
            return searchRec(root.left, value);
        }
        return searchRec(root.right, value);
    }
/**
 * Rysuje drzewo.
 * @return drzewo jako ciąg znaków
 */
    public String draw() {
        StringBuilder result = new StringBuilder();
        drawRec(root, 0, result);
        return result.toString();
    }
    /**
     * Rysuje drzewo przy użyciu rekurencji.
     * @param node - bieżący węzeł
     * @param space - dodatkowa ilość spacji
     * @param result - wynikowy ciąg znaków
     */
    private void drawRec(TreeNode<T> node, int depth, StringBuilder result) {
        if (node == null) return;
        drawRec(node.left, depth + 1, result);
        for (int i = 0; i < depth*4; i++)
            result.append("  "); 
        result.append(node.value);
        result.append("\n \n");
        drawRec(node.right, depth + 1, result);
    }
}